### PR TITLE
initial controller support, moved some things

### PR DESCRIPTION
**Pull request by mardigontoler**

### CODE REVIEW

File: app/matrixController/controller.py [hunk 1]
```diff
@@ -1,24 +1,34 @@
...
 
-        # TODO implement a MatrixArtist and have it handle
-        # drawing to a generic matrixdevice object
-        
-        # TODO spawn a flask server to get animation data
-        # to give to the MatrixArtist
+        self.artist = MatrixArtist(self.device)
+
+
+    def updateData(self, data):
```
Comment (evh111) [inline app/matrixController/controller.py:24]: What do you mean by "not safe on it's own"?

File: app/matrixController/devices/gpioMatrix.py [hunk 2]
```diff
@@ -8,5 +8,24 @@ class GpioMatrix(MatrixDevice):
...
         super().__init__(_M, _N)
 
-    # TODO implement gpio writing
-    
\ No newline at end of file
+    def selectSection(self, section):
+        return
+        
+
+    def writeTopPixel(self, pixel):
```
Comment (evh111) [inline app/matrixController/devices/gpioMatrix.py:15]: What will be the difference in the 'writeBottomPixel' and 'writeTopPixel' functions?

File: app/matrixController/devices/pixel.py [hunk 3]
```diff
@@ -37,8 +50,11 @@ def getAsciimaticsColor(self):
...
             return '- '
-        return '■ '
+        return '¤ '
 
 
     def __repr__(self):
         return str((self.r, self.g, self.b))
+
+    def __eq__(self, other):
+        return (self.r, self.g, self.b) == (other.r, other.g, other.b)
```
Comment (evh111) [inline app/matrixController/devices/pixel.py:60]: I'm just confused with the '(other.r, other.g, other.b)'.

File: static/scripts.js [hunk 4]
```diff
@@ -59,18 +59,26 @@ function clearGrid() {
...
+    let row = []
     for (let j = 0; j < N; j++) {
       cell = canvas.rows[i].cells[j];
       // REMINDER: Modify to restrict this to our pallette
-      animation.frames[0].push(cell.getAttribute('controller-color'));
+      row.push(cell.getAttribute('controller-color'));
+      //animation.frames[0].push(cell.getAttribute('controller-color'));
     }
-  console.log(JSON.stringify(animation, null, 4));
+    animation.frames[0].push(row)
```
Comment (evh111) [inline static/scripts.js:80]: Does .push just append the row to the 'frames' array?

File: aniframe.py [hunk 5]
```diff
@@ -1,13 +1,45 @@
...
-def launchController(controller):
-    controller.begin()
 
 if __name__ == '__main__':
-    controller = Controller(useVirtualMatrix=True)
+    # The general idea is to set up flask's routes while 
+    # giving them access to the controller object.
+    # Then, we start a thread for the controller
+    # to run its loop independently of the flask thread.
+    # the flask routes will use a mutex to make sure
```
Comment (evh111) [inline aniframe.py:15]: I'm glad we are using the mutex lol

File: aniframe.py [hunk 6]
```diff
@@ -1,13 +1,45 @@
...
+        return render_template('index.html')
+
+
+    @flaskApp.route('/animation', methods=['POST'])
+    def animation():
+        updateData(request.get_json(force=True))
+        return '', 200
+
+    # How do we want to set this flag? Using argparse? Click?
+    controller = Controller(useVirtualMatrix=True)
```
Comment (evh111) [inline aniframe.py:40]: Could we use argparse to have the user hit a key on the keyboard if they want the virtualMatrix?

File: app/matrixController/devices/gpioMatrix.py [hunk 7]
```diff
@@ -8,5 +8,24 @@ class GpioMatrix(MatrixDevice):
...
+        return
```
Comment (evh111) [inline app/matrixController/devices/gpioMatrix.py:16]: What will the 'clock' and 'latch' functions do?

File: aniframe.py (hunk 6 above)
Comment (mardigontoler) [reply aniframe.py:40]: argparse would just let us pass in command line arguments when they start up the program.
So they could use `python aniframe.y --virtual` to run it with an optional virtual matrix, etc...

File: app/matrixController/controller.py (hunk 1 above)
Comment (mardigontoler) [reply app/matrixController/controller.py:24]: I mean that calling this without acquiring a lock that protects its data wouldn't be thread-safe. However, the function that _calls_ this function is acquiring the lock; this function doesn't do it on its own.

File: app/matrixController/devices/gpioMatrix.py (hunk 2 above)
Comment (mardigontoler) [reply app/matrixController/devices/gpioMatrix.py:15]: The spec for the matrix indicates that it uses an interlaced signal. As in 1080i vs 1080p, that sort of thing. We write to 2 rows at a time: one in the top half and one in the bottom half. `writeTopPixel` writes to the top row, `writeBottomPixel` writes to the bottom row. Then once we're done with those two rows, we use `latch()`, which tells the matrix that we're done writing pixel data to these rows, and then shift down to wrting to the rows that are below the ones we just wrote. Take another look at the matrix's spec on adafruit, as well as the README.md in the `devices/` directory.

**Pull request merged**